question-answering-using-openai/main.bal

```diff
@@ -0,0 +1,153 @@
+
+// Configure OpenAI client
+final openai:OpenAIClient openaiClient = check new ({auth: {token: openAIToken}});
+
+function countWords(string text) returns int {
+    string[] words = regex:split(text, " ");
+    return words.length();
+}
+
+function cosineSimilarity(decimal[] vector1, decimal[] vector2) returns float {
+    float dotProduct = 0.0;
+    float magnitude1 = 0.0;
+    float magnitude2 = 0.0;
+
+    // Compute dot product and magnitudes
+    foreach int i in 0 ..< vector1.length() {
+        dotProduct += <float>vector1[i] * <float>vector2[i];
+        magnitude1 += (<float>vector1[i]).pow(2);
+        magnitude2 += (<float>vector2[i]).pow(2);
+    }
+
+    // Compute cosine similarity
+    float magnitudeProduct = magnitude1.sqrt() * magnitude2.sqrt();
+    if (magnitudeProduct == 0.0) {
+        return 0.0;
+    }
+    return dotProduct / magnitudeProduct;
+}
+
+function getEmbedding(string text) returns decimal[]|error {
+    openai:CreateEmbeddingRequest input = {
+        input: text,
+        model: "text-embedding-ada-002"
+    };
+    openai:CreateEmbeddingResponse embeddingRes = check openaiClient->/embeddings.post(input);
+    return embeddingRes.data[0].embedding;
+}
+
+function getDocumentSimilarity(string question, map<decimal[]> doc_embeddings) returns [string, float?][]|error {
+    // Get question embedding
+    decimal[] question_embedding = check getEmbedding(question);
+
+    [string, float?][] doc_similarity = [];
+
+    foreach string heading in doc_embeddings.keys() {
+        float similarity = cosineSimilarity(<decimal[]>doc_embeddings[heading], question_embedding);
+        doc_similarity.push([heading, similarity]);
+    }
+
+    [string, float?][] doc_similarity_sorted = from var item in doc_similarity
+        order by item[1] descending
+        select item;
+
+    return doc_similarity_sorted;
+}
+
+function constructPrompt(string question, map<string> documents, map<decimal[]> doc_embeddings) returns string|error {
+    [string, float?][] document_similarity = check getDocumentSimilarity(question, doc_embeddings);
+    string context = "";
+    int contextLen = 0;
+    int maxLen = 1125; // approx equivalence between word and token count
+
+    foreach [string, float?] item in document_similarity {
+        string heading = item[0];
+        string content = <string>documents[heading];
+
+        contextLen += countWords(content);
+        if contextLen > maxLen {
+            break;
+        }
+
+        context += "\n*" + content;
+    }
+
+    string instruction = "Answer the question as truthfully as possible using the provided context, and if the answer is not contained within the text below, say \"I don't know.\"\n\nContext:\n";
+    return string `${instruction} ${context} \n\n Q: ${question} \n A:`;
+}
+
+function generateAnswer(string prompt) returns string|error {
+    openai:CreateCompletionRequest prmt = {
```

Comment:
Move this code to the service it self, for better clarity.

Reply:
addressed in [fa5629f](https://github.com/ballerina-guides/ai-samples/pull/2/commits/fa5629f9ca6bfae66689af4f34151a7883b405d2)